### PR TITLE
Enhanced RHUI script with the changes proposed by RHUI PG

### PR DESCRIPTION
**Pull request by dibaskar**
1)    repo server connectivity validation using domain name instead of IP address. – 1 hour
2)	suggestion to check /etc/hosts file as well along with nameserver entries in /etc/resolv.conf during name resolution check.
3)	Logging of script execution
4)     Disclaimer inclusion

### CODE REVIEW

File: Linux_scripts/RHUI_repo_validation_scripts/README.md [hunk 1]
```diff
@@ -1,19 +1,16 @@
...
+3. If no errors observed, then prints that repo connectivity is successful.
+4. If any errors are observed validates the error with the defined conditions and provides recommendation for the fix.
+
+# Supported OS Images
+
+This version of the script currently supports only Redhat VMs for now(rhe6, rhel7 and 8 non-byos VMs which are deployed from Azure market place image).
+## Usage
+Execute the below commands on Redhat VM for script execution.<br>
+<pre><code class="lang-azurecli" data-interactive="azurecli">mkdir /tmp/rhui && cd /tmp/rhui </code></pre>
+<pre><code class="lang-azurecli" data-interactive="azurecli">wget https://github.com/Azure/azure-support-scripts/archive/refs/heads/master.zip && unzip master.zip 'azure-support-scripts-master/Linux_scripts/RHUI_repo_validation_scripts/*' && rm -f master.zip && cd azure-support-scripts-master/Linux_scripts/RHUI_repo_validation_scripts </code></pre>
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/README.md:15]: I think it's better to publish release and ask customer to take the latest release. The main branch can also have errors/bugs/issues at some point and release will help in taking the correct verified zip.

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py [hunk 2]
```diff
@@ -1,40 +1,53 @@
...
 def redhat():
       ##### checking connectivity of RHEL Repo server IPs #####
 
    def ip_tables_check():
-      reject_rule = subprocess.Popen("iptables -L OUTPUT -v -n | grep -i reject", stdout=subprocess.PIPE, shell=True)
+      reject_rule = subprocess.Popen("iptables -L OUTPUT -v -n | egrep -i 'reject|drop'", stdout=subprocess.PIPE, shell=True)
       rej_rule = str(reject_rule.communicate()[0])
-      if "tcp dpt:443 reject-with icmp-port-unreachable" in rej_rule or "tcp dpt:https reject-with icmp-port-unreachable" in rej_rule or "   reject-with icmp-port-unreachable" in rej_rule:
-         print(rej_rule)
+      if "tcp dpt:443 reject-with icmp-port-unreachable" in rej_rule or "tcp dpt:https reject-with icmp-port-unreachable" in rej_rule or "   reject-with icmp-port-unreachable" in rej_rule or "tcp dpt:443" in rej_rule:
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:21]: make these commands as constants and add them to a list,
iterate that list to check for matching condition

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py [hunk 3]
```diff
@@ -1,99 +1,96 @@
...
    exit()
 
 def msg_yum_repolist_success():
-   print(colors.GREEN + "yum check-update is successful. No errors observed" + colors.ENDC)
+   logging.info(colors.GREEN + "yum check-update is successful. No errors observed" + colors.ENDC)
    exit()
 
 def msg_new_scenario():
-   print(colors.GREEN + "This seems to be a new issue and not the regular one, please engage Azure Linux escalation team for further troubleshooting" + colors.ENDC)
+   logging.info(colors.RED + "This seems to be a new issue and not the regular one, please engage Azure Linux escalation team for further troubleshooting" + colors.ENDC)
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/scenarios.py:50]: add logging.error

Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/scenarios.py:50]: add azlinux escalation contact steps

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py [hunk 4]
```diff
@@ -1,99 +1,96 @@
...
-   print ("\nscenario 2: if '#nslookup rhui-1.microsoft.com' fails then")
-   msg_dns_resolution_error()
-   print ("\nscenario 3: if '#openssl x509 -in /etc/pki/rhui/product/content*.crt -text|grep -E 'Not Before|Not After'' shows 'Not After' date in past then")
-   msg_cert_expiry_error() 
+   logging.info(colors.YELLOW + "\nscenario 2: if '#openssl x509 -in /etc/pki/rhui/product/content*.crt -text|grep -E 'Not Before|Not After'' shows 'Not After' date in past then" + colors.ENDC)
+   msg_cert_expiry_error()
 
 def msg_python_config_issue():
-  print (colors.RED + "seems to be configuration issue with python.Fix python config issues and re-run this script" + colors.ENDC)
+  logging.error(colors.RED + "seems to be configuration issue with python.Fix python config issues and re-run this script" + colors.ENDC)
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/scenarios.py:96]: Nit: add space after full stop "python.Fix"

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py [hunk 5]
```diff
@@ -1,99 +1,96 @@
...
 
 def msg_rhui_package_check_error():
-   print (colors.RED + "rhui-azure package is not available, Install rhui-azure by downloading manually and rerun this script" + colors.ENDC)
+   logging.error(colors.RED + "rhui-azure package is not available, Install rhui-azure by downloading manually and rerun this script" + colors.ENDC)
    exit()
 
 
 def msg_ip_tables_check():
-   print (colors.RED + "Validate the Reject rules in OUTPUT chain which is listed above and please make sure connectivity to RHUI IPs over 443 is not blocked" + colors.ENDC)   
+   logging.error(colors.RED + "Validate the Reject rules in OUTPUT chain which is listed above and please make sure connectivity to RHUI IPs over 443 is not blocked" + colors.ENDC)
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/scenarios.py:35]: mention the RHUI IPs

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py [hunk 6]
```diff
@@ -142,6 +182,8 @@ def os_distro():
 def main():
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:182]: rename it to the init function

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py [hunk 7]
```diff
@@ -75,10 +88,10 @@ def yum_clean_up():
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:88]: except block is missing, add it and capture error in case it caches one

**Review by anujkmaurya — CHANGES_REQUESTED** [review]
First round of review

File: Linux_scripts/RHUI_repo_validation_scripts/README.md (hunk 1 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/README.md:15]: Versioning is applied on the script as per the suggestion.

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py (hunk 2 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/repo_check.py]: Done

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py (hunk 3 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/scenarios.py]: Done

Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/scenarios.py]: Didn't provide azlinux escalation team DL ID as this script will also be executed by customer. But provided suggestion to log support ticket.

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py [hunk 8]
```diff
@@ -1,99 +1,96 @@
...
+   logging.info (colors.GREEN + "RESOLUTION : Please validate your nameserver entries in /etc/resolv.conf file, also make sure that there is no customized entries for RHUI servers in /etc/hosts and then rerun this script" + colors.ENDC)
+   exit()
 
 def msg_https_403_error():
-   print (colors.RED + "\nSeems to be some internal error which is causing the connectivity issue to RHUI server.Please apply the temporary workaround which is mentioned below and see whether it fixes the issue, till our engineering team provides permanent fix" + colors.ENDC)
-   print ("Workaround : Add the below entries in /etc/hosts\n 13.91.47.76  rhui-1.micosoft.com\n 13.91.47.76  rhui-2.micosoft.com\n 13.91.47.76  rhui-3.micosoft.com")
+   logging.error (colors.RED + "\nSeems to be some internal error which is causing the connectivity issue to RHUI server.Please apply the temporary workaround which is mentioned below and see whether it fixes the issue, till our engineering team provides permanent fix" + colors.ENDC)
+   logging.info ("WORkAROUND : Add the below entries in /etc/hosts\n 13.91.47.76  rhui-1.micosoft.com\n 13.91.47.76  rhui-2.micosoft.com\n 13.91.47.76  rhui-3.micosoft.com")
```
[the remark this answers is not in the document]
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/scenarios.py]: Included other IP addresses as well.

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py [hunk 9]
```diff
@@ -1,99 +1,96 @@
...
-    print (colors.GREEN + "Try executing #echo $(. /etc/os-release && echo $VERSION_ID) > /etc/yum/vars/releasever and see whether it fixes the issue" + colors.ENDC)
+    logging.error(colors.RED + "It seems repo is EUS and file /etc/yum/vars/releasever is either not available or have incorrect content" + colors.ENDC)
+    logging.info(colors.GREEN + "RESOLUTION : Try executing #echo $(. /etc/os-release && echo $VERSION_ID) > /etc/yum/vars/releasever and see whether it fixes the issue" + colors.ENDC)
 
 def msg_https_404_possible_cause():
-    print (colors.GREEN + "if repo is non-eus try executing #rm /etc/yum/vars/releasever and see whether it fixes the issue. if repo is eus Try executing #echo $(. /etc/os-release && echo $VERSION_ID) > /etc/yum/vars/releasever and see whether it fixes the issue" +colors.ENDC)
+    logging.info(colors.GREEN + "RESOLUTION : if repo is non-eus try executing #rm /etc/yum/vars/releasever and see whether it fixes the issue. if repo is eus Try executing #echo $(. /etc/os-release && echo $VERSION_ID) > /etc/yum/vars/releasever and see whether it fixes the issue" +colors.ENDC)
```
[the remark this answers is not in the document]
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/scenarios.py:62]: Done as suggested.

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py [hunk 10]
```diff
@@ -1,99 +1,96 @@
...
-   print (" #yum update --disablerepo=* --enablerepo='*microsoft*'")
+   logging.error(colors.RED + "\nSeems RHUI client certificate is expired. rhui-azure package requires an update" + colors.ENDC)
+   logging.info(colors.GREEN + "Fix : Please execute the below command as mentioned in https://docs.microsoft.com/en-us/azure/virtual-machines/workloads/redhat/redhat-rhui for installing the latest version of rhui-azure package" + colors.ENDC)
+   logging.info ("\n #yum update --disablerepo=* --enablerepo='*microsoft*'")
 
 def msg_rhel8_possible_causes():
-   print (colors.RED + "Failed to synchronize cache for repo. Possible scenarios are listed below" + colors.ENDC)  
-   print ("\nscenario 1: Execute '#rpm -qa | grep -i rhui' and checked whether repo is eus or non-eus") 
+   logging.error(colors.RED + "Failed to synchronize cache for repo. Possible scenarios are listed below" + colors.ENDC)
+   logging.info(colors.YELLOW + "\nscenario 1: Execute '#rpm -qa | grep -i rhui' and checked whether repo is eus or non-eus" + colors.ENDC)
```
[the remark this answers is not in the document]
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/scenarios.py:90]: This condition is only applicable for RHEL 8.1 VMs where repo connectivity error code will be generic for all the scenarios and hence our suggestions are also generic.

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/scenarios.py]: Validated the CLI output and it looks good.

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py (hunk 4 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/scenarios.py]: Done

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py (hunk 5 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/scenarios.py]: Done

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py (hunk 6 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/repo_check.py]: Done

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py (hunk 7 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:95]: Added except block wherever missing

**Comment by dibaskar** [comment]
Enhanced the code according to the suggestion provided in PR. Please validate and let us know whether we are good to release this project.

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py [hunk 11]
```diff
@@ -131,6 +150,38 @@ def error_check():
          msg_new_scenario()
    error_check() 
 
+
+def disclaimer():
+   logging.info("\n\n Disclaimer : This script helps in detecting RHUI server connectivity issues only for Redhat-6,7 and 8 Pay as you go images deployed from Azure market place.Upon execution this script does not modify anything on the VM instead it will provide suggestion for fix based on the error. Always, Please download the latest version of the script from git hub page which will help in effective troubleshooting and also have bugs addressed.\n")
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:155]: just check and correct the disclaimer message: eg: place.Upon -> place. Upon
Always, Please -> Always
git hub -> GitHub page (add link too)

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py [hunk 12]
```diff
@@ -1,40 +1,58 @@
...
 
    def ip_tables_check():
-      reject_rule = subprocess.Popen("iptables -L OUTPUT -v -n | grep -i reject", stdout=subprocess.PIPE, shell=True)
+      reject_rule = subprocess.Popen("iptables -L OUTPUT -v -n | egrep -i 'reject|drop'", stdout=subprocess.PIPE, shell=True)
       rej_rule = str(reject_rule.communicate()[0])
-      if "tcp dpt:443 reject-with icmp-port-unreachable" in rej_rule or "tcp dpt:https reject-with icmp-port-unreachable" in rej_rule or "   reject-with icmp-port-unreachable" in rej_rule:
-         print(rej_rule)
-         msg_ip_tables_check()
+      error_list = ["tcp dpt:443 reject-with icmp-port-unreachable", "tcp dpt:https reject-with icmp-port-unreachable", "   reject-with icmp-port-unreachable", "tcp dpt:443"]
+      for var in error_list:
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:25]: for error in error_list:

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py [hunk 13]
```diff
@@ -1,40 +1,58 @@
...
+   ##### rhui connectivity check ########
 
    def rhui_ip_conn_check():
-      print ("\nchecking connectivity of RHUI repo server IPs")
-      ip = ["13.91.47.76", "40.85.190.91", "52.187.75.218", "52.174.163.213", "52.237.203.198"]
+      logging.info("checking connectivity of RHUI repo server IPs \n")
+      servers = ["rhui-1.microsoft.com", "rhui-2.microsoft.com", "rhui-3.microsoft.com"]
       port = "443"
-      for var in ip:
+      for var in servers:
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:54]: for server in servers

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py [hunk 14]
```diff
@@ -67,6 +85,8 @@ def yum_clean_up():
         stdout, stderr = yum_clean.communicate()
         if "Another app is currently holding the yum lock" in stderr.decode() or "Waiting for process with" in stderr.decode():
            msg_multiple_yum_process()
+     except:
+         logging.error("Issue happened while executing yum clean all command or while gathering yum processes list. Fix the issue and rerun the script")
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:89]: Following Error is encountered while executing...

Here we are talking about the yum errors and suggestions, right?

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py [hunk 15]
```diff
@@ -87,6 +107,8 @@ def yum_repo_check():
          my_timer.start()
          stdout, stderr = yum.communicate()
          return stderr
+      except:
+         logging.error("There seems to be some issue while processing yum check-update command output, rerun the script after fixing the issue")
```
Comment (anujkmaurya) [inline Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:111]: what is the customer supposed to do?

**Review by anujkmaurya — CHANGES_REQUESTED** [review]

File: Linux_scripts/RHUI_repo_validation_scripts/scenarios.py [hunk 16]
```diff
@@ -1,99 +1,96 @@
...
+   logging.error (colors.RED + "\nSeems to be some internal error which is causing the connectivity issue to RHUI server.Please apply the temporary workaround which is mentioned below and see whether it fixes the issue, till our engineering team provides permanent fix" + colors.ENDC)
+   logging.info ("\nWORKAROUND:\n#########For Azure Global add the below entries in /etc/hosts file#######\n 13.91.47.76  rhui-1.micosoft.com rhui-2.micosoft.com rhui-3.micosoft.com\n 40.85.190.91  rhui-1.micosoft.com rhui-2.micosoft.com rhui-3.micosoft.com\n 52.187.75.218  rhui-1.micosoft.com rhui-2.micosoft.com rhui-3.micosoft.com\n 52.174.163.213  rhui-1.micosoft.com rhui-2.micosoft.com rhui-3.micosoft.com\n 52.237.203.198  rhui-1.micosoft.com rhui-2.micosoft.com rhui-3.micosoft.com\n\n######### For Azure Azure US Government add the below entries in /etc/hosts file#######\n 13.72.186.193  rhui-1.micosoft.com rhui-2.micosoft.com rhui-3.micosoft.com\n 13.72.14.155  rhui-1.micosoft.com rhui-2.micosoft.com rhui-3.micosoft.com\n 52.244.249.194  rhui-1.micosoft.com rhui-2.micosoft.com rhui-3.micosoft.com")
```
[the remark this answers is not in the document]
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/scenarios.py:81]: Done

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py (hunk 13 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:54]: Done

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py (hunk 11 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:155]: This is Done.

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py (hunk 12 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:25]: Modified the variable name

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py (hunk 14 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:89]: If there is issue is getting yum process list due to performance issue or ps command got hung then this message will be printed.

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py (hunk 15 above)
Comment (dibaskar) [reply Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:111]: There might be some chances like yum check-update is hung due to performance issues or any other config issues, CX need to manually execute this command to check whats happening.

**Review by anujkmaurya — APPROVED** [review]

File: Linux_scripts/RHUI_repo_validation_scripts/repo_check.py (hunk 15 above)
Comment (anujkmaurya) [reply Linux_scripts/RHUI_repo_validation_scripts/repo_check.py:111]: please instruct the customer to manually run the given command here

**Pull request merged**